File: 24_SwapNodesinPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode temp = new ListNode(0);
        temp.next = head;
        ListNode l1 = temp;
        ListNode l2 = head;
        while(l2 != null && l2.next != null) {
        	ListNode nextPositio = l2.next.next;
        	l1.next = l2.next;
        	l2.next.next = l2;
        	l2.next = nextPositio;
        	l1 = l2;
        	l2 = l2.next;
        }
        return temp.next;
    }
}

File: c++/24_SwapNodesinPairs.cpp
//sol1
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return head;
        ListNode* headNext = head->next;
        head->next = swapPairs(headNext->next);
        headNext->next = head;
        return headNext;
    }
};
//sol2
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        for (ListNode* p = dummy; p->next && p->next->next;) {
            auto a = p->next, b = a->next;
            p->next = b;
            a->next = b->next;
            b->next = a;
            p = a;
        }
        return dummy->next;
    }
};
